torch/nn/parallel/distributed.py

```diff
@@ -633,27 +633,60 @@ def _register_comm_hook(self, state: object, hook: callable):
                              c10d reducer would call this hook and use the tensors returned
                              by the Future and copy grads to individual parameters.
 
+                             We also provide an API called ``get_future`` to retrieve a
+                             Future associated with the completion of c10d.ProcessGroupNCCL.work.
+                             Note that this API will return a toch._C.Future which is an
+                             internal type and should be used with caution.
+
         .. warning ::
             DDP communication hook can only be registered once and should be registered
             before calling backward.
 
         .. warning ::
-            The torch.futures.Future object that hook returns should contain a result that
-            has the same shape with the tensors inside GradBucket bucket.
+            The Future object that hook returns should contain a result that has the same
+            shape with the tensors inside grad bucket.
+
+        .. warning ::
+            DDP communication hook does not support single process multiple device mode.
+            Gradbucket tensors should consist of only a single tensor.
 
         .. warning ::
             DDP communication hook is experimental and subject to change.
 
         Example::
             Below is an example of a noop hook that returns back the same tensors:
 
-            >>> ddp._register_comm_hook(state = None, hook = noop)
-
-            >>> def noop(state: object, bucket: dist.GradBucket): -> torch.futures.Future
+            >>> def noop(state: object, bucket: dist._GradBucket): -> torch.futures.Future
             >>>     fut = torch.futures.Future()
             >>>     fut.set_result(bucket.get_tensors())
             >>>     return fut
 
+            >>> ddp._register_comm_hook(state = None, hook = noop)
+
+        Example::
+            Below is an example of a simple allreduce hook.
+
+            >>> def allreduce(state: object, bucket: dist._GradBucket): -> torch._C.Future
+            >>>     work = process_group.allreduce(bucket.get_tensors())
+            >>>     return work.get_future()
+
+            >>> ddp._register_comm_hook(state = None, hook = allreduce)
+
+        Example::
+            Below is an example of a Parallel SGD algorithm where gradients are encoded before
+            allreduce, and then decoded after allreduce.
+
+            >>> def encode_and_decode(state: object, bucket: dist._GradBucket): -> torch._C.Future
```

Comment:
Does `.then()` return `FutureNCCL` or `torch.futures.Future`? I assume it's the latter? In this case, we the return type should be `torch.futures.Future`? I guess we only need to when `torch._C.Future` as the return type when the hook function directly returns `FutureNCCL`.

Reply:
I tried the following:
```c++
        def allreduce_then_hook(state: object, bucket: dist._GradBucket) -> torch._C.Future:
            tensors = bucket.get_tensors()#[t / self.world_size for t in bucket.get_tensors()]
            fut = process_group.allreduce(tensors).get_future()

            def mult(fut):
                # Multiply result by 10.
                return [10 * t for t in fut.wait()]

            def div(fut):
                print(type(fut))
                # Divide result by 2.
                return [0.5 * t for t in fut.wait()]

            return fut.then(mult).then(div)
```
`print(type(fut))` prints `torch._C.Future`. I think any future that has something to do with C++ no matter its either `at::ivalue::Future ` or `FutureNCCL` will be this type in Python. cc'ed @pritamdamania87

Reply:
It is true that all these future types are some wrapper/derived type of `torch._C.Future`, but we were only intended to use that to refer to `FutureNCCL`. The original thought is that we have the following mapping:

`torch._C.Future` -> `FutureNCCL`
`torch.futures.Future` -> all other futures

Since I cannot explain this clearly between us during the past few rounds of discussions, I guess this might be even harder to convey accurately to users. Does the following make more sense to you?

1. Create a new pybind type that exposes `FutureNCCL` as `torch.distributed.FutureNCCL` to Python.
2. Explicitly mark the return type of `get_future()` as `torch.distributed.FutureNCCL`.
3. Use `torch.futures.Future` to refer to all other future types.
4. In the API doc of `FutureNCCL`, explain the difference between `torch.distributed.FutureNCCL` and `torch.futures.Future`

Reply:
@mrshenli Please see the latest version. I hope I covered everything based on our discussion offline.